### DISCUSSION TITLE
cpp/basics/type_conversions

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#52] [Announcements]
# cpp/basics/type_conversions

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/basics/type_conversions.html

**Comment by Whaleman55** [comment]
const_cast对常量进行转换不是未定义的行为吗？

**Comment by huangjiawei128** [comment]
补充——static_cast和C中的强制类型转换的区别：
static_cast：从 Derived* 到 Base* 的上行转换和从 Base* 到 Derived* 的下行转换是安全的，并且通过编译时检查。然而，试图将 Base* 转换为与之无关的类 Unrelated* 会导致编译错误。
强制类型转换：将 Base* 转换为与之无关的类 Unrelated* 并不会导致编译错误。

**Comment by yysczl** [comment]
为什么需要虚拟函数
虚表（vtable）：当一个类中至少有一个虚拟函数时，编译器会为这个类生成一个虚表（vtable）。虚表是一个指针数组，包含了该类及其派生类的虚拟函数的地址。每个对象都会有一个指向其类的虚表的指针（通常称为 vptr），这个指针使得在运行时能够找到对象的实际类型。
类型安全转换：dynamic_cast 通过检查对象的 vptr 来判断对象的实际类型。如果没有虚拟函数，编译器不会为类生成 vtable，也就无法提供 RTTI 信息。因此，dynamic_cast 无法工作，因为它没有足够的信息来确定对象的真实类型。
